src/box/lua/key_def.h

```diff
@@ -56,6 +63,65 @@ luaT_push_key_def(struct lua_State *L, const struct key_def *key_def);
 struct key_def *
 luaT_check_key_def(struct lua_State *L, int idx);
 
+/**
+ * Extract key from tuple by given key definition and return
+ * tuple representing this key.
+ * Push the new key tuple as cdata to a LUA stack on success.
+ * Raise error otherwise.
+ */
+int
+luaT_key_def_extract_key(struct lua_State *L, struct key_def *key_def,
+			 int idx_t);
+
+/**
+ * Compare tuples using the key definition.
+ * Push 0  if key_fields(tuple_a) == key_fields(tuple_b)
+ *      <0 if key_fields(tuple_a) < key_fields(tuple_b)
+ *      >0 if key_fields(tuple_a) > key_fields(tuple_b)
+ * integer to a LUA stack on success.
+ * Raise error otherwise.
+ */
+int
+luaT_key_def_compare(struct lua_State *L, struct key_def *key_def, int idx_a,
+		     int idx_b);
```

Comment:
The signautre of this and other functions added by this patch look weird, not Lua-style, because some arguments are passed as C pointers while others are passed as indexes in Lua stack. Wouldn't it be better to pass the `key_def` via the Lua stack as well? Also, having that many indexes is confusing IMO. I'd change the signature of this function to

```C
int
luaT_key_def_compare(struct lua_State *L, int idx);
```

and make it compare the top two values at the Lua stack with the `key_def` located at `idx`. This would be consistent with `lua_settable`, for example. If the values that needs to be compared are located not at the top of the Lua stack, one can push them to the top with `lua_pushvalue`.

Reply:
NB: `lua_settable()` pop two elements from a *top of the stack*. I don't remember any Lua/C API function that has access to `idx+1` or `idx-1` element (relative to the passed `idx`).

Reply:
> and make it compare the top two values at the Lua stack with the key_def located at idx.

This is exactly what the initial patch looked like: https://github.com/tarantool/tarantool/pull/8356#discussion_r1111338354

Reply:
Discussed f2f. Agreed to pass arguments in the Lua way, as suggested in the first comment of this thread.

Also, agreed that we need to rename `luaT_check_key_def` to `luaT_is_key_def`, because `check` functions usually throw on error.

Reply:
Done